### DISCUSSION TITLE
Pull actual supercharging fees from account API

### DISCUSSION THREAD

**Discussion by dthiery** [teslamate-org/teslamate#1831] [Ideas]
Actual supercharger fees for a session can be pulled from the Tesla API by hitting https://www.tesla.com/teslaaccount/charging/api/history?vin=5YJ3XXXXXXXXXX. It would be a cool feature to pull supercharging actual costs from the acct API after a charging session completes instead of relying on geofenced cost numbers.

**Comment by dhruvarora2** [comment]
Would be nice to have this!

**Comment by bikeymouse** [comment]
Second best thing would be if there was a (community based?) list of SuperCharge locations and costs per kWh per country that you could easily import into Teslamate as Geofences. 
In general an import/export function for Geofences would be nice, so we could share Geofence.

**Reply by prisnagka** [reply]
I agree, but it's also the case of different cost depending on the time of day. Using the api directly would eliminate that problem. (I wonder if the api includes any extra costs connected to parking too long after completing a charge as well?

**Comment by prisnagka** [comment]
> Actual supercharger fees for a session can be pulled from the Tesla API by hitting https://www.tesla.com/teslaaccount/charging/api/history?vin=5YJ3XXXXXXXXXX.

This would seriously be such an awesome addition that I'm surprised that it's not already in there. Does anyone know if there are any specific restrictions posed by Tesla that would make this extra hard or impossible to implement?

**Reply by cwanja** [reply]
I should not have deleted my comment, but rather edited. 

The only issue I could see with this would be lining up charging sessions - you could guesstimate by start and end time. But if the API is delayed in either end (likely the owners over the charging one), I could see it getting wonky. It _seems_ technically feasible, but fully admitting I do not know the under the covers architecture that @adriankumpf has developed.

**Comment by bouvrais** [comment]
I am a bit confused. I just installed Teslamate. Went to a SC today and expected the charge fee to show up in Teslamate but it does not. The charge session does show up but the cost column is empty. 
I understand that this discussion is about pulling the actual fee from the API, but what is the current (expected) behaviour? Should Teslamate infer the cost from the sesssion duration? Something else? Or is the behaviour I am seeing (i.e. nothing) by design?

**Reply by cwanja** [reply]
TeslaMate showing nothing for the cost is the expected, current, working behavior. You would need to manually add the cost to the charge session. This discussion wants to bring it in automatically.

**Reply by bikeymouse** [reply]
If you make this SuC or any other charge point a Point of Interest you can add the cost per Kwh so next charges will be calculated automatically based on that.

**Reply by bouvrais** [reply]
Thanks for the quick replies @cwanja, @bikeymouse. I can now see that clicking the "cost" cell on the charge dashboard for a particular charge session brings me to a form where I can input the actual cost for that particular session.
I had indeed to create a geofence for that particular SC (and name it appropriately) for it to be picked up as a supercharger session since its default name does not include "Supercharger"... But it works as intended!
I do concur that pulling the charge session info from the API would be a elegant solution as it would likely kill to birds with one stone: not only could one get the detailed costs, but it would also allow to identify a particular charging session as a SC session without inferring from the name.

**Comment by prisnagka** [comment]
Good suggestion but with prices varying based on time etc, a better approach is still to pull the data via an api. Adding 10-15 SoC’s as poi’s will be cumbersome to maintain.  

Cheers,
/Peter

**Reply by bikeymouse** [reply]
Yes, of course. Just wanted to point a new user to the option that is allready available in the current version.

Btw: I’m not aware of time based pricing for Tesla SuC’s in Europe at least, so maintaining the prices for your 5-10 “favorite” SuC’s seems doable. 
Bigger problem is that what the car is reporting as the added Kwh is less then what the charger actually delivered, due to losses.
So getting the/your actual charge costs from an API will always be more accurate.

**Reply by prisnagka** [reply]
Just got an email from Tesla where they introduce time based pricing - lower prices nighttime when the utilisation is lower. 

We could really use this type of feature if it’s possible to fit in.

**Comment by mat1990dj** [comment]
TeslaFI has already implemented this and it's adding a 5m window for time differences

**Comment by ooottttooo** [comment]
Any news on this idea? It would still be a neat feature. If I could, I would program it :-)

**Comment by jpizquierdo** [comment]
> Actual supercharger fees for a session can be pulled from the Tesla API by hitting https://www.tesla.com/teslaaccount/charging/api/history?vin=5YJ3XXXXXXXXXX. It would be a cool feature to pull supercharging actual costs from the acct API after a charging session completes instead of relying on geofenced cost numbers.

It would be awesome

**Comment by niklaswa** [comment]
I would also appreciate such a feature.

I also pulled the current endpoint for the charging history the Tesla App is using (looks almost the same as the account API, but maybe that can help someone) :)

```
POST https://akamai-apigateway-charging-ownership.tesla.com/graphql?vin=<VIN>&operationName=getChargingHistoryV2
```

<details>

<summary>
GraphQL Query/Variables
</summary>

```graphql
query getChargingHistoryV2(
  $pageNumber: Int!
  $sortBy: String
  $sortOrder: SortByEnum
  $latestSession: Boolean
) {
  me {
    charging {
      historyV2(
        pageNumber: $pageNumber
        sortBy: $sortBy
        sortOrder: $sortOrder
        latestSession: $latestSession
      ) {
        data {
          ...SparkHistoryItemFragment
        }
        totalResults
        hasMoreData
        pageNumber
      }
    }
  }
}

fragment SparkHistoryItemFragment on SparkHistoryItem {
  countryCode
  programType
  billingType
  vin
  isMsp
  credit {
    distance
    distanceUnit
  }
  chargingPackage {
    distance
    distanceUnit
    energyApplied
  }
  chargingVoucher {
    voucherValue
  }
  invoices {
    fileName
    contentId
    invoiceType
  }
  chargeSessionId
  siteLocationName
  chargeStartDateTime
  chargeStopDateTime
  unlatchDateTime
  fees {
    ...SparkHistoryFeeFragment
  }
  vehicleMakeType
  sessionId
  surveyCompleted
  surveyType
  postId
  cabinetId
  din
  isDcEnforced
  siteAmenities
  siteEntryLocation {
    latitude
    longitude
  }
  siteAddress {
    ...SparkAddressFragment
  }
  sessionSource
  additionalNotes {
    left
    right
  }
}

fragment SparkHistoryFeeFragment on SparkHistoryFee {
  sessionFeeId
  feeType
  payorUid
  amountDue
  currencyCode
  pricingType
  usageBase
  usageTier1
  usageTier2
  usageTier3
  usageTier4
  rateBase
  rateTier1
  rateTier2
  rateTier3
  rateTier4
  totalTier1
  totalTier2
  totalTier3
  totalTier4
  uom
  isPaid
  uid
  totalBase
  totalDue
  netDue
  status
  showPeriods
  periods {
    ...SparkHistoryFeePeriodsFragment
  }
}

fragment SparkHistoryFeePeriodsFragment on SparkHistoryFeePeriods {
  sessionFeePeriodId
  startDateTime
  stopDateTime
  actualQuantity
  rate
}

fragment SparkAddressFragment on SparkAddressType {
  street
  streetNumber
  city
  district
  state
  countryCode
  country
  postalCode
}
```
Variables:
```json
{
    "sortBy": "start_datetime",
    "sortOrder": "DESC",
    "pageNumber": 1
}
```
</details>

<details>
<summary>
Response looks like this:
</summary>

```json
{
  "countryCode": "NL",
  "programType": "PTSCH",
  "billingType": null,
  "vin": "LRW3E7ET7RC000000",
  "isMsp": false,
  "credit": null,
  "chargingPackage": null,
  "chargingVoucher": null,
  "chargeSessionId": "00000000-0000-0000-0000-000000000000",
  "siteLocationName": "Heerlen, Netherlands",
  "chargeStartDateTime": "2024-10-25T20:34:50+02:00",
  "chargeStopDateTime": "2024-10-25T21:08:29+02:00",
  "unlatchDateTime": "2024-10-25T21:08:33+02:00",
  "vehicleMakeType": "TESLA",
  "sessionId": 000000000,
  "surveyCompleted": false,
  "surveyType": "CSAT",
  "postId": "3",
  "cabinetId": "XX0000000000X0",
  "din": "STST-SM-00000",
  "isDcEnforced": false,
  "siteAmenities": [
    "AMENITIES_RESTROOMS",
    "AMENITIES_CAFE",
    "AMENITIES_LODGING",
    "AMENITIES_RESTAURANT",
    "AMENITIES_TWENTY_FOUR_HOUR"
  ],
  "sessionSource": null,
  "additionalNotes": [
    
  ],
  "siteEntryLocation": {
    "latitude": 50.892974,
    "longitude": 5.955635
  },
  "siteAddress": {
    "street": "Terworm",
    "streetNumber": "10",
    "city": "Heerlen",
    "district": null,
    "state": null,
    "countryCode": "NL",
    "country": "Netherlands",
    "postalCode": "6411 RV"
  },
  "invoices": [
    {
      "fileName": "000000000_DE-DE.pdf",
      "contentId": "00000000-0000-0000-0000-000000000000",
      "invoiceType": "IMMEDIATE"
    }
  ],
  "fees": [
    {
      "sessionFeeId": 000000000,
      "feeType": "CHARGING",
      "payorUid": null,
      "amountDue": null,
      "currencyCode": "EUR",
      "pricingType": "PAYMENT",
      "usageBase": 48.595,
      "usageTier1": 9,
      "usageTier2": 23,
      "usageTier3": null,
      "usageTier4": null,
      "rateBase": 0.23,
      "rateTier1": 0,
      "rateTier2": 0,
      "rateTier3": null,
      "rateTier4": null,
      "totalTier1": 0,
      "totalTier2": 0,
      "totalTier3": 0,
      "totalTier4": 0,
      "uom": "kWh",
      "isPaid": true,
      "uid": null,
      "totalBase": 11.17,
      "totalDue": 11.17,
      "netDue": null,
      "status": "PAID",
      "showPeriods": null,
      "periods": null
    },
    {
      "sessionFeeId": 000000000,
      "feeType": "PARKING",
      "payorUid": null,
      "amountDue": null,
      "currencyCode": "EUR",
      "pricingType": "NO_CHARGE",
      "usageBase": 0,
      "usageTier1": 0,
      "usageTier2": 0,
      "usageTier3": null,
      "usageTier4": null,
      "rateBase": 0,
      "rateTier1": 0,
      "rateTier2": 0,
      "rateTier3": null,
      "rateTier4": null,
      "totalTier1": 0,
      "totalTier2": 0,
      "totalTier3": 0,
      "totalTier4": 0,
      "uom": "min",
      "isPaid": true,
      "uid": null,
      "totalBase": 0,
      "totalDue": 0,
      "netDue": null,
      "status": "PAID",
      "showPeriods": null,
      "periods": null
    }
  ]
}
```
</details>

**Comment by giterdone509** [comment]
Would love this

**Comment by florianfish** [comment]
I vote for this awesome feature! <3

**Comment by stauffenberg2020** [comment]
@JakobLichterfeld , Can you please let the interested parties know if this requested feature is something that can be looked or has if it been completely discarded?

**Comment by JakobLichterfeld** [comment]
See: https://github.com/teslamate-org/teslamate/issues/3266#issuecomment-2219732903
